### ISSUE TITLE
Lombok plugin

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#525]
Provides annotations for getters/setters/equals/hashcode:  https://projectlombok.org/

I suggest we incorporate it. Should help reduce a good bit of code. Any objections?

**Comment by gaborbernat** [comment]
I don't like it that much. It involves byte code mangling and only works with a random IDE if you install a plugin for it.

**Comment by DanVanAtta** [comment]
Does the bytecode manging/manipulation cause any practical problems? I agree it's not the best, but I have trouble seeing it as a 'con' since it doesn't cause any practical problems that I am aware of.

I agree the one-time IDE set up is a small pain, but it's one time. Certainly a 'con', but once documented I have trouble seeing it as that major of a problem.

Meanwhile there would be a lot of benefit. We use getters/setters/ data type objects a lot. Using the annotations we would be able to get rid of at least 2000 lines of code, probably more on the order of 3-6  thousand lines. That's a lot of code that would just go away! "The best code is the code you do not have to write"

Here are some counts of getters/setters/equals/hashcode/toString:
```
dan@dan-xps:~/work/triplea$ find . -name "*java" -exec grep "public.*get.*{" {} \; | wc -l
2318
dan@dan-xps:~/work/triplea$ find . -name "*java" -exec grep "public.*set.*{" {} \; | wc -l
1372
dan@dan-xps:~/work/triplea$ find . -name "*java" -exec grep "public.*toString.*{" {} \; | wc -l
126
dan@dan-xps:~/work/triplea$ find . -name "*java" -exec grep "public.*equals.*{" {} \; | wc -l
36
dan@dan-xps:~/work/triplea$ find . -name "*java" -exec grep "public.*hashCode.*{" {} \; | wc -l
30
```

**Comment by gaborbernat** [comment]
My main worry is that it removes straight forward code, that rarely shelters bugs. And it does it by byte code manipulation which at first may seem like magic to a not so experienced Java developer, thus increasing the entry barrier. If the number of settters getters bother you why not change them to public fields, and bypass the setter/getters that way. With the new IDEs, it's easy to find the read/write points of a field.

**Comment by DanVanAtta** [comment]
equals and hashcode can shelter bugs quite easily. Also if it's trivial code, why not just replace the 3 lines with an annotation? Then it's clear that it's a drop dead simple setter/getter with no extra magic.

IMO new developers are not going to be pained by this too much. Sure bytecode insertion will seem like magic, but so will reflection. I don't really see us as increasing the barrier to entry very much if at all. Removing the complicated equals methods would be lowering it, but fundamentally I think it's very easy for anyone to figure out what a "@Getter" annotation would do.

I'll point out that the setter/getter methods are required by reflection in some cases, so we can't just change variables to public. Also, it's far too much work to try and convert everything to an immutable data pattern that would allow for public data members. So the combination of having to update so many constructors, and the requirements of reflection to have no-arg constructors and getter/setter methods defeats the suggestion to instead just use an immutable data pattern.

**Comment by gaborbernat** [comment]
I see your point. I'm neither in favor or against it. Maybe @Ron and @veryqn should also voice their thoughts.

**Comment by ron-murhammer** [comment]
@DanVanAtta I think there is too much invisible magic and things are fragile with Lombok. Also would increase the learning curve for most new contributors which is already pretty high given the code complexity. I'd be a little more open to Google AutoValue but not even sure I'd want to adopt that at this point.

**Comment by DanVanAtta** [comment]
How so more fragile? I'd suggest the opposite since we surely have a lot of bugs in the equals implementations, plus overhead of ensuring we keep all properties correctly wired together in hashcode and equals.

**Comment by DanVanAtta** [comment]
Second question, should we be optimizing for theoretical new contributors, or should we be optimizing for ourselves and the code in it's existing state? Wouldn't also less code, help new contributors? Again, is there that much hidden magic or uncertainty about what a "@Getter" annotation would do, or seeing "@Hashcode" at the top of a class. Second point to that, a new person would simply have to emulate the "@Hashcode" at the top of a class rather than learning how to write a proper hash code implementation, which is not trivial. These points indicate it would be easier for the theoretical new contributor, not more difficult. For us, shaving off 10-20% no the code base and not having to worry about implementing equals and hashcode is valuable.

**Comment by DanVanAtta** [comment]
@ron-murhammer  Google AutoValue  is interesting, looks like a good way to create builders and immutable classes. The magic behind a builder / immutable class with google autovalue I think is actually more sophisticated than that of @getter/@setter/@equals annotations.

**Comment by ron-murhammer** [comment]
This slide summarizes the cons pretty well: https://docs.google.com/presentation/d/14u_h-lMn7f1rXE1nDiLX0azS3IkgjGl5uxp5jGJ75RE/edit#slide=id.g2a5e9c4a8_0104. Really depends you weight those cons against the benefits it would bring.

I think we should strive to keep the new contributor barrier as low as possible. Most developers are gonna know core Java but there are so many libraries that most probably won't have ever used Lombok. It also doesn't seem incredibly popular and the magic makes understanding it more difficult than many of the other libraries we have adopted.

Less code isn't always better if its hard to follow what's going on. Writing the entire code base as a single class and method would probably also reduce at least 10-20% but I would never advocate doing it :)

**Comment by DanVanAtta** [comment]
I agree we should strive to keep the new contributor barrier low, but that elides the points made about how keeping things simple helps. New contributors are utterly lost today. I'm sorry to say that is a lost battle. Not using tools/frameworks that would simplify things does not help new contributors, it only serves to maintain the current mess. That mess then further deters new contributors. I have literally gotten the response of "I have no idea where to start" from potential new contributors.

Taking 5 minutes to read what the annotations do on either page does not take long. In fact, new contributors might be drawn to the project by us using interesting/useful frameworks, whether that be lombok, dagger, or google autovalue. 

I wonder how good of a presentation that was. l

>Pros:
Maximally concise value classes.

- Misses, simplification from having to maintain equals/hashcode, plus removal of bugs since those are non-trivial.

Cons:
The inserted code is invisible. This makes for a poor experience with tools, such as debuggers and code explorers.
- code explorers? really? IntelliJ or eclipse handle it well with a 5 minute step of a plugin installation. Debugger could be bad for the time while the pointer is in the auto-generated code, but likely people are not stepping into a getter method. For an auto-generated equals method, similar, likely not stepping into it, and unit tests is a better way to ensure correctness in the long run. So this is con is complete crap.

- The compiler hacks are non-standard and fragile.
I would like to see this proven or pointed out in specific.. I can cite examples where Lombok has been deployed at scale without issue. If it is fragile, we will find out pretty soon, in which worst case whomever investigated this will hit a dead end.

- "In our view, your code is no longer really Java (Lombok is very "extralinguistic")"
Perhaps this just needs to hear the talk. The bullet point is complete BS as it stands.

**Comment by DanVanAtta** [comment]
I'd rather focus on dependency injection first. So I'm willing to let this go.

If we switch to a native game save, rather than reflection, lombok could prove very useful. If someone wanted to volunteer to lombok large sections of code, I think that would be a good thing too.

**Comment by DanVanAtta** [comment]
actually, this is stupid.. the top 10 google hits for lombok are pretty rave:

http://keaplogik.blogspot.com/2013/04/java-developers-need-to-be-using-lombok.html
https://gualtierotesta.wordpress.com/2014/03/03/tutorial-using-lombok-to-reduce-boilerplate-code-in-java/
http://www.springfuse.com/2009/10/18/project-lombok-must-have-in-your-java-toolbox.html
http://jnb.ociweb.com/jnb/jnbJan2010.html (Reducing Boilerplate Code with Project Lombok, by Michael Kimberlin, Senior Software Engineer  Object Computing, Inc. (OCI))
http://stackoverflow.com/questions/3852091/is-it-safe-to-use-project-lombok 

@ron-murhammer , what is your evidence/reasoning for saying that lombok is not popular?

**Comment by DanVanAtta** [comment]
@ron-murhammer  - Any real reasons to not use this plugin?

**Comment by ron-murhammer** [comment]
@DanVanAtta I'd prefer to focus on something like dependency injection first as well as I think there are more benefits to that.

Well most of the links related to lombok are old (several years) and I've never been involved with a project that's used it. It also has a very minimal website and is written by a bunch of individual authors. I'm not sure if you have a better popularity measure.

I think I've stated several and whether or not you accept them is up to you. @gaborbernat also didn't appear to be in love with it either.

**Comment by DanVanAtta** [comment]
It's not a matter of me accepting them, it's more a matter of having valid reasons. The reasoning of: ""In our view, your code is no longer really Java (Lombok is very "extralinguistic")"   is complete crap. An actual reason I could consider and accept. But it's just been so far unsubstantiated opinions.

So far the only substantive argument I"m hearing is that this would make it hard for new developers. I've countered with many more reasons why it would actually help new dev's.

**Comment by DanVanAtta** [comment]
Minimal website does not mean much. The conversations about it are recent and ongoing: https://groups.google.com/forum/#!forum/project-lombok. I don't think you can say it's not popular just as much as you can say it is.

Googling for google Guava, most of the first page hits are a few years old. A stack overflow article from 2011 is number 5. IF the first page of "java project lombok" was a littany of "this project sucks", then I could see your point. But instead it's quite the opposite, the first page is virtually nothing but praise, and that continues on subsequent search pages.

**Comment by DanVanAtta** [comment]
Pending actionable/concrete objections/concerns, we would need to do the following: 
- add documentation for devs to get set up on eclipse + intelliJ
- verify reflection issues/concerns. Are there any implications for save game compatibility?
- functional testing, verify all play modes work fine, PBEM, PBF, single and multiplayer.

**Comment by ron-murhammer** [comment]
@DanVanAtta I really don't want to argue over this. Bottom line is that I'm against it. If you can convince @veqryn that's its a good idea then I'm fine moving forward with it.

**Comment by DanVanAtta** [comment]
Let's do this, down the road, let's see how this looks in the small. Perhaps after seeing actual benefit, maybe you won't be so opposed.

I have a couple of objections to this discussion:
1. You cherry-picked a clearly biased slide, and this is the second time you've cherry-picked a very biased argument, and then called that a day. Shifting and subjective arguments against are simply not convincing nor terribly useful.
2. I can't tell how much of this is the fear of the unknown, vs an actual objection. You were very opposed to Dagger, but now you're interested in adding it in it seems...
3. For plugins where there is only one choice, if it's useful, I think we should start using them. Whoever adds it in has plenty of work. It's virtually pointless now to have these a priori hypothetical conversations, that resemble acts of congress.   Before it made more sense to get approval since it was another lib to add, an increase in the game DL. But we just saved 200MB off the game DL, and we import libraries at build time. There is very little to not use frameworks that will be of value/help. When there are multiple frameworks that accomplish the same goal, then I see a lot of merit in having a discussion about which one to pick.
4. I don't think any one person should have veto powers over how we update/do things. 
5. There is far too much opinion subjectivity thrown around. Why should we convince anyone it's a good idea / bad idea, it should be very clear. From that we simply need to look at the pro's + cons. The cons have been layed out: 
- byte code manipulation, nobody is a fan of it
- requires IDE plugin
Pros:
- gobs of boiler plate code disappears.

**Comment by DanVanAtta** [comment]
I was curious about the popularity question. Google trends is actually a good way to measure that. Interestingly enough, lombok turns out to be more popular than dagger, and google autovalue does not look to be taking off yet:

![lombok_autovalue_dagger](https://cloud.githubusercontent.com/assets/12397753/13867269/96e35b22-ec7a-11e5-879a-1920517cfb83.png)


Of some more interest, here are some trends of the 'major' frameworks, lombok interestingly is not completely dwarfed on the this graph:
![major_frameworks_vs_lombok](https://cloud.githubusercontent.com/assets/12397753/13867338/3cbccde4-ec7b-11e5-9d54-eb71f09d29cc.png)

**Comment by veqryn** [comment]
I think I would like this at some point down the road, but I feel like there are much higher priorities right now, including a few that I feel might be pre-requisites to this, such as removing every trace of reflection in the project.
I am generally not a fan of "magic", so if we did make use of this, I would argue we should do it only for the largest use cases (getters, setters, equals, hashcode, tostring, and nothing else).
How about punting on this for 4 months?
